### PR TITLE
Convert more traditional for-loops to use index_range() helper function

### PR DESCRIPTION
**Pull request by jwpeterson**
This is fairly granular to make bisection a bit easier in case any of the testing fails. I can squash it down later if desired.

### CODE REVIEW

**Comment by pbauman** [comment]
GRINS failure:

<pre>
src/partitioning/libmesh_devel_la-partitioner.lo
/femputer/pbauman/civet_build_testing_root/libMesh/src/parallel/request.C: In function ‘void libMesh::Parallel::wait(std::vector<libMesh::Parallel::Request>&)’:
/femputer/pbauman/civet_build_testing_root/libMesh/src/parallel/request.C:215:15: error: declaration of ‘request’ shadows a global declaration [-Werror=shadow]
   for (auto & request : r)
               ^~~~~~~
In file included from /femputer/pbauman/civet_build_testing_root/libMesh/src/parallel/request.C:19:0:
./include/libmesh/request.h:48:21: note: shadowed declaration is here
 typedef MPI_Request request;
</pre>

**Comment by jwpeterson** [comment]
Thanks @pbauman. Same problem with the complex numbers build. I guess I need to add `-Werror -Wshadow` to my `CXXFLAGS`... 🙄

**Comment by roystgnr** [comment]
Just configure with --enable-paranoid-warnings --enable-werror so you get everything we enforce.

**Comment by jwpeterson** [comment]
I fixed all the paranoid warnings that showed up for clang, but the one pointed out by @pbauman wasn't even there... if there are more failures, I'll compile with GCC and fix those as well.

**Comment by jwpeterson** [comment]
> I'll compile with GCC and fix those as well.

So at least with GCC 6.2.0 it's not complaining about any more `-Wshadow` errors for me... I'll see what the build boxes say. And I can actually see the femputer output since I'm working from home so I should be able to eventually fix things by trial and error...

**Comment by jwpeterson** [comment]
Squashed quite a bit and rebased, if this still passes I will go ahead and merge.

File: src/geom/elem_cutter.C [hunk 1]
```diff
@@ -283,11 +283,11 @@ void ElemCutter::cut_2D (const Elem & elem,
   _inside_elem.clear();
   _outside_elem.clear();
 
-  for (const auto & elem : _inside_mesh_2D->element_ptr_range())
-    _inside_elem.push_back (elem);
```
Comment (roystgnr) [inline src/geom/elem_cutter.C]: How did this *not* trigger -Wshadow before?

File: src/partitioning/subdomain_partitioner.C [hunk 2]
```diff
@@ -56,14 +57,12 @@ void SubdomainPartitioner::_do_partition (MeshBase & mesh,
...
   // subdomains in question, and pass it to the internal Partitioner.
-  for (std::size_t c=0; c<chunks.size(); ++c)
-    {
-      MeshBase::element_iterator
-        it = mesh.active_subdomain_set_elements_begin(chunks[c]),
-        end = mesh.active_subdomain_set_elements_end(chunks[c]);
-
-      _internal_partitioner->partition_range(mesh, it, end, n);
-    }
+  for (auto c : index_range(chunks))
```
Comment (roystgnr) [inline src/partitioning/subdomain_partitioner.C]: Do we want an index_range here?  Looks like a range for over chunks would work.

File: src/reduced_basis/rb_scm_evaluation.C [hunk 3]
```diff
@@ -529,7 +529,7 @@ void RBSCMEvaluation::legacy_read_offline_data_from_files(const std::string & di
 
   // Resize C_J based on C_J_stability_vector and Q_a
   C_J.resize( C_J_length );
-  for (std::size_t i=0; i<C_J.size(); i++)
+  for (auto i : index_range(C_J))
```
Comment (roystgnr) [inline src/reduced_basis/rb_scm_evaluation.C]: This could also just be a plain range-for, right?

**Comment by roystgnr** [comment]
:+1: regardless of whether you fix those quibbles

**Comment by roystgnr** [comment]
Merging; this is a solid improvement and there's no telling when John's going to have time to address my nitpicks.

File: src/geom/elem_cutter.C (hunk 1 above)
Comment (jwpeterson) [reply src/geom/elem_cutter.C]: Code is only active in `--disable-strict-lgpl` builds, not sure if we have one of those that *also* has `-Werror -Wshadow`

File: src/numerics/distributed_vector.C [hunk 4]
```diff
@@ -234,13 +231,19 @@ void DistributedVector<T>::add (const NumericVector<T> & v)
...
 
 template <typename T>
-void DistributedVector<T>::add (const T a, const NumericVector<T> & v)
+void DistributedVector<T>::add (const T a, const NumericVector<T> & v_in)
 {
   libmesh_assert (this->initialized());
   libmesh_assert_equal_to (_values.size(), _local_size);
   libmesh_assert_equal_to ((_last_local_index - _first_local_index), _local_size);
 
-  add(a, v);
```
[the remark this answers is not in the document]
Comment (jwpeterson) [reply src/numerics/distributed_vector.C]: If someone had called it, yeah. It was found by the `-Winfinite-recursion` warning.

File: src/numerics/sum_shell_matrix.C [hunk 5]
```diff
@@ -30,7 +31,7 @@ numeric_index_type SumShellMatrix<T>::m () const
   libmesh_assert(!matrices.empty());
   const numeric_index_type n_rows = matrices[0]->m();
 #ifndef NDEBUG
-  for (std::size_t i=1; i<matrices.size(); ++i)
+  for (auto i : index_range(matrices))
```
[the remark this answers is not in the document]
Comment (jwpeterson) [reply src/numerics/sum_shell_matrix.C]: Yeah.

File: src/reduced_basis/rb_scm_evaluation.C (hunk 3 above)
Comment (jwpeterson) [reply src/reduced_basis/rb_scm_evaluation.C]: Yep, I will change that one.

File: src/partitioning/subdomain_partitioner.C (hunk 2 above)
Comment (jwpeterson) [reply src/partitioning/subdomain_partitioner.C]: Yes, will change.

**Pull request merged**